pg_es_fdw/_es_query.py

```diff
@@ -43,7 +44,7 @@ def _base_qual_to_es(col, op, value, column_map=None):
         return {"bool": {"must_not": {"term": {col: value}}}}
 
     if op == "~~":
-        return {"match": {col: value.replace("%", "*")}}
+        return {"wildcard": {col: value.replace("%", "*")}}
```

Comment:
Ideally this would handle underscores (single-character match) and escapes (e.g. `\%` matches `%`), so something like:

  - `%` (when not preceded by a single `\`) becomes `*`
  - `_` (when not preceded by a single `\`) becomes `?`
  - `\%` becomes `%` (since it has no special meaning in ES wildcards)
  - `\_` becomes `_` (same)

It does become complex though. I'd be happy with us returning false positives (since PG will recheck and filter them out) but I don't know how to process the input LIKE pattern to get to a pattern that definitely over-matches instead of possibly under-matching (e.g. right now the `x LIKE 'a_ple'` won't match `apple` AFAIU)

Reply:
Right, I think [this](https://github.com/splitgraph/postgres-elasticsearch-fdw/commit/ff95446620a4c13df28d2db33e3cd297e51dd54e) should do it, let me add a couple of tests in the engine repo.

Reply:
Now that I think of it, I need to handle case like `\\%`(becomes `\*`), and `\\\%` (becomes `\%`?).

Reply:
* `\\%` in PG patterns means "backslash + any characters", so it'd become `\\*`
  * `\\\%` means "backslash + `%`", so it'd become `\\%` (since `%` has no special meaning)

Reply:
Got it, [this should now](https://github.com/splitgraph/postgres-elasticsearch-fdw/commit/f3dd1ea7ef619cd94d7488bb80913149c41b4286) be covered as well. [Here}(https://github.com/splitgraph/splitgraph/blob/multicorn-agg-pushdown-v2-cu-1z461e4/test/splitgraph/commands/test_es_fdw_queries.py#L36-L79) are the corresponding tests.

Reply:
Can you also test / check that this escapes special ES pattern characters, e.g.:

  * `*` becomes `\*`
  * `?` becomes `\?`

Reply:
Good catch, added a [fix](https://github.com/splitgraph/postgres-elasticsearch-fdw/commit/4dfffd72553afd17be40ba19f98e7e3a8e072b2d) and a [test](https://github.com/splitgraph/splitgraph/commit/d4a805f985db11017c0b0f1d937ba496e5b457e4) for these cases too.